### PR TITLE
`Sniffer::wait_for_message_type`

### PR DESCRIPTION
**Pull request by plebhash**
close https://github.com/stratum-mining/stratum/issues/1282

this PR adds a new method `Sniffer::wait_for_message_type` which takes two parameters:
- `message_direction: MessageDirection`
- `message_type: u8`

the function will return execution as soon as there's some expected message type on some `Sniffer` FIFO (either downstream or upstream)

this can be used to hold the execution of the tokio runtime of some Integration Test

it's useful when we don't know beforehand how long it could take to e.g.: find a share

### CODE REVIEW

**Comment by jbesraa** [comment]
I tried to use it here https://github.com/stratum-mining/stratum/pull/1262/commits/e478edb4714f5b3cf50eb5fdf691ea5fd596573a but sometimes it goes forever so added this https://github.com/stratum-mining/stratum/pull/1262/commits/482ce929f86370c289ed2f9d223d40697d36b139

**Comment by plebhash** [comment]
> I tried to use it here [e478edb](https://github.com/stratum-mining/stratum/commit/e478edb4714f5b3cf50eb5fdf691ea5fd596573a) but sometimes it goes forever so added this [482ce92](https://github.com/stratum-mining/stratum/commit/482ce929f86370c289ed2f9d223d40697d36b139)

current theory is that CPU miner is buggy

**Comment by plebhash** [comment]
> > I tried to use it here [e478edb](https://github.com/stratum-mining/stratum/commit/e478edb4714f5b3cf50eb5fdf691ea5fd596573a) but sometimes it goes forever so added this [482ce92](https://github.com/stratum-mining/stratum/commit/482ce929f86370c289ed2f9d223d40697d36b139)
> 
> current theory is that CPU miner is buggy

after some investigation, @GitGab19 and I concluded this is miscalibrated vardiff on tProxy

CPU miner can find shares without problems, but it takes a while before it sends one to tProxy that meets the diff criteria for propagation towards the pool (which would trigger the sniffer)

so I added two new commits to `jbesraa:2024-11-14-rewrite-tproxy-test` branch of #1262: 
- 2c11ff43b4e74c401989b4ec53ece99fcfb0966a "recalibrating" tProxy vardiff
- e572d4d738ec3920fef027f207d35391cd2cab8c making the timeout more conservatively long (only for extreme cases, which should ideally never happen)

these commits allowed me to run `cargo t translation_proxy -- --nocapture` multiple times without the timeout.

however, the test will fail due to unwrap on CPU miner, but the panic error messages vary.

Here's some variations I observed:
```
thread 'translation_proxy' panicked at tests-integration/tests/common/mod.rs:149:45:
called `Result::unwrap()` on an `Err` value: JsonRpc(Transport(SocketError(Os { code: 61, kind: ConnectionRefused, message: "Connection refused" })))
thread 'async-std/runtime' panicked at /Users/plebhash/develop/stratum/roles/test-utils/mining-device-sv1/src/client.rs:114:70:
called `Result::unwrap()` on an `Err` value: RecvError
test translation_proxy ... FAILED
```

which seems just some unrelated bug on the logic that handles TCP on the CPU miner

**Comment by plebhash** [comment]
@jbesraa do you ACK this PR?

if yes we should merge it and keep investigating under #1262

**Review by jbesraa — APPROVED** [review]
ACK

File: roles/tests-integration/tests/common/sniffer.rs [hunk 1]
```diff
@@ -418,6 +418,37 @@ impl Sniffer {
...
+                MessageDirection::ToUpstream => {
+                    self.messages_from_downstream.has_message_type(message_type)
+                }
+            };
+
+            // ready to unblock test runtime
+            if has_message_type {
+                return;
+            }
+
```
[the remark this answers is not in the document]
Comment (jbesraa) [reply roles/tests-integration/tests/common/sniffer.rs:444]: I would avoid adding any sleep unless it is really necessary

[the remark this answers is not in the document]
Comment (jbesraa) [reply roles/tests-integration/tests/common/sniffer.rs:444]: Shouldnt the lock acquisition be dropped once we return?
Or are you saying suggesting to add a sleep after each loop?

[the remark this answers is not in the document]
Comment (plebhash) [reply roles/tests-integration/tests/common/sniffer.rs:444]: AFAIU the suggestion is to sleep on every loop cycle, which sounds reasonable as an optimization of async programming

[the remark this answers is not in the document]
Comment (plebhash) [reply roles/tests-integration/tests/common/sniffer.rs:444]: > I would just drop the timeout to something in seconds and leave this as is.

timeout is a solution for a different problem, it has nothing to do with async lock acquisition

> we already have too many sleep all over testing

this is async sleep, it doesn't block anything

**Pull request merged**